design-proposals/klip-43-timestamp-data-type-support.md

```diff
@@ -0,0 +1,194 @@
+
+**Author**: @jzaralim | 
+**Release Target**: 0.15 | 
+**Status**: In Discussion | 
+**Discussion**: https://github.com/confluentinc/ksql/pull/6649
+
+**tl;dr:** _Add support for TIMESTAMP column types in ksqlDB. This will allow users to easily migrate
+time data from other databases without having to convert column types, as well as simplify time data
+manipulation_
+
+## Motivation and background
+
+With Connect integration, KSQL processes data from various databases. Most other databases
+have a TIMESTAMP or DATETIME type, which KSQL currently converts to BIGINT. This makes sinking to
+databases that don't implicitly cast long values to timestamps much more complicated - the user would
+have to set up their connector to convert these columns back to TIMESTAMP. Supporting TIMESTAMP types
+would make moving time data between KSQL and other data sources/sinks smoother and less error-prone.
+
+Adding a TIMESTAMP type also simplifies time data manipulation. For example, currently
+if a user wants to extract the month of a timestamp, they would either have to parse a string or do a
+lot of math. Having time data in a dedicated data type allows for a lot of new UDFs.
+
+## What is in scope
+
+* Add TIMESTAMP type to KSQL
+* Support TIMESTAMP arithmetic and comparisons
+* Support TIMESTAMP usage in STRUCT, MAP and ARRAY
+* Serialization and de-serialization of TIMESTAMPs to Avro, JSON, Protobuf and Delimited formats
+* Update existing built-in functions to use the TIMESTAMP data type
+* Casting between TIMESTAMP and BIGINT
+
+## What is not in scope
+* Changing the ROWTIME data type. We will eventually want this to happen, but that is a separate
+discussion.
+* Support for dates before Unix epoch - this is not supported by Kafka. There is a [KIP](https://cwiki.apache.org/confluence/display/KAFKA/KIP-228+Negative+record+timestamp+support)
+for negative timestamps, but until that is implemented, KSQL can only support positive timestamps.
+* DATE and TIME types - because TIMESTAMPs represent a point in time, DATE and TIME types
+would be useful if a user wants to represent a less specific time. These would be added after
+TIMESTAMP is implemented though.
+* Allow window units (HOUR, DAY etc) to be used in timestamp functions.
+
+## Public APIS
+
+The TIMESTAMP data type will store a point in time after Unix epoch without timezone information.
+The syntax is as follows:
+
+```roomsql
+CREATE STREAM stream_name (time TIMESTAMP, COL2 STRING) AS ...
+CREATE TABLE table_name (col1 STRUCT<field TIMESTAMP>) AS ...
+```
+
+TIMESTAMPS will be displayed in console as strings in ISO-8601 format:
+
+```roomsql
+> SELECT time FROM stream_name EMIT CHANGES;
+
++------------------------+
+|time                    |
++------------------------+
+|1994-11-05T13:15:30Z    |
+```
+
+TIMESTAMPS can be represented by milliseconds from Unix epoch or by using conversion functions:
+
+```roomsql
+INSERT INTO stream_name VALUES (1605927509166);
```

Comment:
There's a concern I have about the implicit cast here. Some systems (including Mysql, Hive, and Impala) use timestamp as seconds, but Java and other internal formats (i.e. Avro) use timestamp as milliseconds. Why are we choosing milliseconds instead of the common seconds approach? Could users get confused about the value here? Should we prevent this implicit cast instead?
```
mysql> select unix_timestamp('2020-11-30 13:25:00');
+---------------------------------------+
| unix_timestamp('2020-11-30 13:25:00') |
+---------------------------------------+
|                            1606764300 |
+---------------------------------------+
1 row in set (0.00 sec)
```

Reply:
The main reason is because Kafka represents timestamps as milliseconds, and a lot of what we have now is based off that (ROWTIME representation as BIGINT, timestamps being converted to long types) so implicitly casting to milliseconds would help with compatibility. However, it seems that most other DBs use a seconds representation, so a new user could get confused by this. We should explain this clearly in the docs if we decide to implicitly cast.